File: dart/blockchain/lib/consensus/models/protocol_settings.dart
import 'package:blockchain/crypto/impl/kes_product.dart';
import 'package:fixnum/fixnum.dart';
import 'package:rational/rational.dart';

class ProtocolSettings {
  final Rational fEffective;
  final int vrfLddCutoff;
  final int vrfPrecision;
  final Rational vrfBaselineDifficulty;
  final Rational vrfAmpltitude;
  final int chainSelectionKLookback;
  final Duration slotDuration;
  final int operationalPeriodsPerEpoch;
  final int kesKeyHours;
  final int kesKeyMinutes;

  ProtocolSettings(
      {required this.fEffective,
      required this.vrfLddCutoff,
      required this.vrfPrecision,
      required this.vrfBaselineDifficulty,
      required this.vrfAmpltitude,
      required this.chainSelectionKLookback,
      required this.slotDuration,
      required this.operationalPeriodsPerEpoch,
      required this.kesKeyHours,
      required this.kesKeyMinutes});

  factory ProtocolSettings.fromMap(Map<String, String> map) => ProtocolSettings(
      fEffective: _parseRational(map["f-effective"]!),
      vrfLddCutoff: int.parse(map["vrf-ldd-cutoff"]!),
      vrfPrecision: int.parse(map["vrf-precision"]!),
      vrfBaselineDifficulty: _parseRational(map["vrf-baseline-difficulty"]!),
      vrfAmpltitude: _parseRational(map["vrf-amplitude"]!),
      chainSelectionKLookback: int.parse(map["chain-selection-k-lookback"]!),
      slotDuration: Duration(milliseconds: int.parse(map["slot-duration-ms"]!)),
      operationalPeriodsPerEpoch:
          int.parse(map["operational-periods-per-epoch"]!),
      kesKeyHours: int.parse(map["kes-key-hours"]!),
      kesKeyMinutes: int.parse(map["kes-key-minutes"]!));

  ProtocolSettings mergeFromMap(Map<String, String> map) => ProtocolSettings(
        fEffective: map.containsKey("f-effective")
            ? _parseRational(map["f-effective"]!)
            : fEffective,
        vrfLddCutoff: map.containsKey("vrf-ldd-cutoff")
            ? int.parse(map["vrf-ldd-cutoff"]!)
            : vrfLddCutoff,
        vrfPrecision: map.containsKey("vrf-precision")
            ? int.parse(map["vrf-precision"]!)
            : vrfPrecision,
        vrfBaselineDifficulty: map.containsKey("vrf-baseline-difficulty")
            ? _parseRational(map["vrf-baseline-difficulty"]!)
            : vrfBaselineDifficulty,
        vrfAmpltitude: map.containsKey("vrf-amplitude")
            ? _parseRational(map["vrf-amplitude"]!)
            : vrfAmpltitude,
        chainSelectionKLookback: map.containsKey("chain-selection-k-lookback")
            ? int.parse(map["chain-selection-k-lookback"]!)
            : chainSelectionKLookback,
        slotDuration: map.containsKey("slot-duration-ms")
            ? Duration(milliseconds: int.parse(map["slot-duration-ms"]!))
            : slotDuration,
        operationalPeriodsPerEpoch:
            map.containsKey("operational-periods-per-epoch")
                ? int.parse(map["operational-periods-per-epoch"]!)
                : operationalPeriodsPerEpoch,
        kesKeyHours: map.containsKey("kes-key-hours")
            ? int.parse(map["kes-key-hours"]!)
            : kesKeyHours,
        kesKeyMinutes: map.containsKey("kes-key-minutes")
            ? int.parse(map["kes-key-minutes"]!)
            : kesKeyMinutes,
      );

  static const defaultAsMap = {
    "f-effective": "3/25",
    "vrf-ldd-cutoff": "15",
    "vrf-precision": "40",
    "vrf-baseline-difficulty": "1/20",
    "vrf-amplitude": "1/2",
    "chain-selection-k-lookback": "81", // 5184
    "slot-duration-ms": "1000",
    "forward-biased-slot-window": "50",
    "operational-periods-per-epoch": "25",
    "kes-key-hours": "9",
    "kes-key-minutes": "9"
  };

  @override
  String toString() =>
      "ProtocolSettings(fEffective=$fEffective, vrfLddCutoff=$vrfLddCutoff, vrfPrecision=$vrfPrecision, vrfBaselineDifficulty=$vrfBaselineDifficulty, vrfAmplitude=$vrfAmpltitude, kLookback=$chainSelectionKLookback, slotDuration=${slotDuration.inMilliseconds}ms, operationalPeriodsPerEpoch=$operationalPeriodsPerEpoch, kesHeight=(${kesKeyHours}, ${kesKeyMinutes}), operationalPeriodLength=$operationalPeriodLength, epochLength=$epochLength)";

  static final ProtocolSettings defaultSettings =
      ProtocolSettings.fromMap(defaultAsMap);

  int get chainSelectionSWindow =>
      (Rational(BigInt.from(chainSelectionKLookback), BigInt.from(4)) *
              fEffective.inverse)
          .round()
          .toInt();

  Int64 get epochLength =>
      Int64((Rational(BigInt.from(chainSelectionKLookback)) *
              fEffective.inverse *
              Rational(BigInt.from(3)))
          .round()
          .toInt());

  Int64 get operationalPeriodLength =>
      epochLength ~/ operationalPeriodsPerEpoch;

  TreeHeight get kesTreeHeight => TreeHeight(kesKeyHours, kesKeyMinutes);
}

_parseRational(String value) {
  final split = value.split("/");
  if (split.length == 1)
    return Rational(BigInt.parse(split[0]));
  else
    return Rational(BigInt.parse(split[0]), BigInt.parse(split[1]));
}

File: scala/core/src/main/scala/blockchain/consensus/ProtocolSettings.scala
package blockchain.consensus

import blockchain.utility.Ratio
import cats.Show
import cats.implicits.showInterpolator

import scala.concurrent.duration.{FiniteDuration, given}

case class ProtocolSettings(
    fEffective: Ratio,
    vrfLddCutoff: Int,
    vrfPrecision: Int,
    vrfBaselineDifficulty: Ratio,
    vrfAmplitude: Ratio,
    vrfSlotGap: Int,
    chainSelectionKLookback: Int,
    slotDuration: FiniteDuration,
    operationalPeriodsPerEpoch: Int,
    kesKeyHours: Int,
    kesKeyMinutes: Int
):

  val chainSelectionSWindow: Int =
    (Ratio(chainSelectionKLookback, 4) * fEffective.inverse).round.toInt

  val epochLength: Int =
    (Ratio(chainSelectionKLookback * 3) * fEffective.inverse).round.toInt

  val operationalPeriodLength: Int =
    epochLength / operationalPeriodsPerEpoch

  val kesTreeHeight: TreeHeight =
    TreeHeight(kesKeyHours, kesKeyMinutes)

  require(epochLength % 3L == 0, s"Epoch length=$epochLength must be divisible by 3")
  require(
    epochLength % operationalPeriodsPerEpoch == 0,
    s"Epoch length=$epochLength must be divisible by $operationalPeriodsPerEpoch"
  )

  override def toString: String =
    show"ProtocolSettings(" +
      show"fEffective=$fEffective," +
      show" vrfLddCutoff=$vrfLddCutoff," +
      show" vrfPrecision=$vrfPrecision," +
      show" vrfBaselineDifficulty=$vrfBaselineDifficulty," +
      show" vrfAmplitude=$vrfAmplitude," +
      show" vrfSlotGap=$vrfSlotGap," +
      show" kLookback=$chainSelectionKLookback," +
      show" slotDuration=$slotDuration," +
      show" operationalPeriodsPerEpoch=$operationalPeriodsPerEpoch," +
      show" kesHeight=($kesKeyHours, $kesKeyMinutes)," +
      show" operationalPeriodLength=$operationalPeriodLength slots," +
      show" epochLength=$epochLength slots" +
      show")"

  def merge(map: Map[String, String]): ProtocolSettings =
    map.foldLeft(this)(_.withSetting.apply.tupled(_))

  def withSetting(name: String, value: String): ProtocolSettings =
    name match {
      case "f-effective"                   => copy(fEffective = parseRational(value))
      case "vrf-ldd-cutoff"                => copy(vrfLddCutoff = value.toInt)
      case "vrf-precision"                 => copy(vrfPrecision = value.toInt)
      case "vrf-baseline-difficulty"       => copy(vrfBaselineDifficulty = parseRational(value))
      case "vrf-amplitude"                 => copy(vrfAmplitude = parseRational(value))
      case "vrf-slot-gap"                  => copy(vrfSlotGap = value.toInt)
      case "chain-selection-k-lookback"    => copy(chainSelectionKLookback = value.toInt)
      case "slot-duration-ms"              => copy(slotDuration = value.toLong.milli)
      case "operational-periods-per-epoch" => copy(operationalPeriodsPerEpoch = value.toInt)
      case "kes-key-hours"                 => copy(kesKeyHours = value.toInt)
      case "kes-key-minutes"               => copy(kesKeyMinutes = value.toInt)
      case _                               => throw IllegalArgumentException(name)
    }
  private def parseRational(value: String): Ratio =
    value.split('/') match {
      case Array(numerator)              => Ratio(BigInt(numerator))
      case Array(numerator, denominator) => Ratio(BigInt(numerator), BigInt(denominator))
      case _                             => throw IllegalArgumentException(value)
    }
  def toMap: Map[String, String] =
    Map(
      "f-effective" -> fEffective.toString,
      "vrf-ldd-cutoff" -> vrfLddCutoff.toString,
      "vrf-precision" -> vrfPrecision.toString,
      "vrf-baseline-difficulty" -> vrfBaselineDifficulty.toString,
      "vrf-amplitude" -> vrfAmplitude.toString,
      "vrf-slot-gap" -> vrfSlotGap.toString,
      "chain-selection-k-lookback" -> chainSelectionKLookback.toString,
      "slot-duration-ms" -> slotDuration.toMillis.toString,
      "operational-periods-per-epoch" -> operationalPeriodsPerEpoch.toString,
      "kes-key-hours" -> kesKeyHours.toString,
      "kes-key-minutes" -> kesKeyMinutes.toString
    )

object ProtocolSettings:

  val Default: ProtocolSettings = ProtocolSettings(
    fEffective = Ratio(3, 25),
    vrfLddCutoff = 18,
    vrfPrecision = 40,
    vrfBaselineDifficulty = Ratio(1, 20),
    vrfAmplitude = Ratio(1, 2),
    vrfSlotGap = 1,
    chainSelectionKLookback = 576,
    slotDuration = 3000.milli,
    operationalPeriodsPerEpoch = 12,
    kesKeyHours = 5,
    kesKeyMinutes = 5
  )

  given Show[ProtocolSettings] = Show.fromToString

case class TreeHeight(hours: Int, minutes: Int) {
  def asTuple: (Int, Int) = (hours, minutes)
}
